internal/webhooks/clusterclass.go

```diff
@@ -168,6 +176,216 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
+	// Note: The index is used to improve the error recording below.
+	newVars, clusterClassVariablesIndex := getClusterClassVariablesMapWithReverseIndex(new.Spec.Variables)
+
+	// Compute the diff between old and new ClusterClassVariables.
+	varsDiff := getClusterClassVariablesForValidation(oldVars, newVars)
+
+	errorInfo := errorAggregator{
+		altered:  variableClusterReferences{},
+		added:    variableClusterReferences{},
+		removed:  variableClusterReferences{},
+		required: variableClusterReferences{},
+	}
+	clusterList := []string{}
+	// Validate the variable values on each Cluster ensuring they are still compatible with the new ClusterClass.
+	for _, cluster := range clusters {
+		clusterList = append(clusterList, cluster.Name)
+		for _, clusterVar := range cluster.Spec.Topology.Variables {
+			// Add Cluster Variable entry in clusterVariableReferences to track where it is in use.
+			tracker.add(clusterVar.Name, cluster.Name)
+
+			// 1) Error if a variable with a schema altered in the update is no longer valid on the Cluster.
+			if alteredVar, ok := varsDiff.altered[clusterVar.Name]; ok {
+				if errs := variables.ValidateClusterVariable(clusterVar.DeepCopy(), alteredVar, field.NewPath("")); len(errs) > 0 {
+					errorInfo.altered.add(alteredVar.Name, cluster.Name)
+				}
+			}
+
+			// 2) Error if a variable removed in the update is still in use in some Clusters.
+			if _, ok := varsDiff.removed[clusterVar.Name]; ok {
+				errorInfo.removed.add(clusterVar.Name, cluster.Name)
+			}
+
+			// 3) Error if a variable has been added in the update check is no longer valid on the Cluster.
+			// NOTE: This can't occur in normal circumstances as a variable must be defined in a ClusterClass in order to be introduced in
+			// a Cluster. This check may catch errors in cases involving broken Clusters.
+			if addedVars, ok := varsDiff.added[clusterVar.Name]; ok {
+				if errs := variables.ValidateClusterVariable(clusterVar.DeepCopy(), addedVars, field.NewPath("")); len(errs) > 0 {
+					errorInfo.added.add(clusterVar.Name, cluster.Name)
+				}
+			}
+		}
+	}
+
+	// 4) Error if a required variable is not defined in every cluster using the ClusterClass.
+	for varName := range varsDiff.required {
+		clustersMissingVariable := clustersWithoutVar(clusterList, tracker.get(varName))
+		if len(clustersMissingVariable) > 0 {
+			errorInfo.required.add(varName, clustersMissingVariable...)
+		}
+	}
+	// Aggregate the errors from the validation checks and return one error message of each type for each variable with a list of violating clusters.
+
+	return aggregateErrors(errorInfo, clusterClassVariablesIndex, path)
+}
+
+// errorAggregator used to aggregate information about errors during variable comparison.
+// TODO(killianmuldoon): Make the errorAggregator generic so it can be used in cases where a single error aggregates errors related to many objects.
```

Comment:
I have a hard time understanding how the current implementation can be evolved to something generic as it seems to be very coupled to the ideas of:
* altered / added / removed / required variables
* field errors

Seems like we have to abstract away:
* specific "error states" (altered, added,...) or implement the generic aggregator only for one of them and then use it multiple times here
* the error creation (including error message generation and field error generation)

Sounds doable, but I'm not sure how useful / reusable the result would be. But something for a potential other PR/issue, right?

Reply:
Yeah - I think there's a lot to think about on this one - I'll create an issue for it.

Reply:
Something to think about is to not use specific fields to signal the action taken on it, for example today we have `altered`, `added`, `removed`, `required`. These could be kinds of errors which we might enumerate somewhere else and then have a single aggregator struct that when the error is added works on the enumeration rather than going list by list.

Does that make sense? Happy to expand further

Reply:
I've updated this now, but in a way that doesn't really serve to make it more generic (just removes all the extra maps) I'd like to punt on this and tackle the more generic error handling in a future PR WDYT?